### PR TITLE
✨ [Carousel]: Initial amp-base-carousel Bento component

### PR DESCRIPTION
**Pull request by caroqliu**
[Demo](https://storage.googleapis.com/amp-test-website-1/amp_dist_77815/test/manual/amp-base-carousel/basic.amp.html)

This PR does the following:
- Uses externs to define BaseCarousel and associated component prop types
- Registers the `amp-base-carousel` custom element as `1.0` using the underlying BaseCarousel Preact component
- Exposes AMP actions and events available on `amp-base-carousel:0.1`
- Add unit tests for `amp-base-carousel`

Notes: 
- This PR exposes three AMP actions: `prev`, `next`, and `goToSlide(index)`. In the case of the first two, a prop `setAdvance` is used to pass the smooth scroll advance method from the Preact layer to the AMP layer. In the case of the last, `mutateProps` is used to set a new slide index causing a render to that slide position without smooth scrolling. This is OK because, when looping, the scroll direction between the current slide and the target slide may appear confusing due to their dynamic placements when looping is enabled. If preferred it is possible to have the first two methods also employ this method of advancing to avoid having to pass the `advance` from child to the parent, at the cost of no longer smooth scrolling.
- ~The `type.js` pattern for defining component props appears to cause a type mismatch with the JSX syntax -- namely that the `PreactDef.FunctionalComponent` employs an argument of `JsonObject`. When this argument was changed to i.e. `ScrollerProps`, the following change was required to silence this type error:
`return <Scroller {...props}> {children} </Scroller>` -> `return {Scroller({...props, children})}`~
- ~Use of `...rest` in `BaseCarousel` appears to cause lint error even though it is fine elsewhere: `Cannot read property 'type' of undefined`~

### CODE REVIEW

File: extensions/amp-base-carousel/1.0/base-carousel.js [hunk 1]
```diff
@@ -16,61 +16,82 @@
...
-        advance={() => advance(-1)}
-      />
-      <Arrow
-        customArrow={arrowNext}
-        dir={1}
-        disabled={disableForDir(1)}
-        advance={() => advance(1)}
-      />
+    <div>
+      {Scroller({
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.js:75]: JSX?

File: extensions/amp-base-carousel/1.0/base-carousel.type.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * @typedef {{
+ *   children: !Array<PreactDef.Renderable>,
+ *   ignoreProgrammaticScroll: {current: (boolean|null)},
+ *   loop: (boolean|undefined),
+ *   restingIndex: number,
+ *   setRestingIndex: (function(number):undefined),
+ * }}
+ */
+var ScrollerProps;
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.type.js:43]: The extern namespace is global. So, unfortunately, we need to be very expansive in our names. E.g. `BaseCarouselScrollerProps` or such. But pinging @jridgewell to confirm. I think closure conflated two things in externs "global namespace" and "type external to compiler".

File: src/preact/base-element.js [hunk 3]
```diff
@@ -341,16 +341,21 @@ function collectProps(Ctor, element, defaultProps) {
 
       const {single, name, props: slotProps = {}} = def;
 
-      // TBD: assign keys, reuse slots, etc.
+      // TBD: assign keys, etc.
```
Comment (dvoytenko) [inline src/preact/base-element.js:344]: Nit: I still hope to reuse slots :)

File: extensions/amp-base-carousel/1.0/base-carousel.js (hunk 1 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.js]: I restored to JSX. To clarify the change originally was meant as an illustrative one for what it takes to silence the type errors as they stand now.

File: src/preact/base-element.js (hunk 3 above)
Comment (caroqliu) [reply src/preact/base-element.js:344]: Ah, I had removed this because I modified this to reuse rather than overwrite existing `slot` attributes, but I can restore the comment if it means to do something different.

File: extensions/amp-base-carousel/1.0/base-carousel.type.js (hunk 2 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.type.js]: Renamed.

**Comment by caroqliu** [comment]
@jridgewell Any advice on why [this lint error](https://travis-ci.org/github/ampproject/amphtml/jobs/708866864) is surfacing? 
```
Message:
    Cannot read property 'type' of undefined
Occurred while linting /home/travis/build/ampproject/amphtml/extensions/amp-base-carousel/1.0/base-carousel.js:30
Details:
    currentNode: [object Object]
    domainEmitter: [object Object]
    domainThrown: false
```

File: extensions/amp-base-carousel/1.0/base-carousel.js [hunk 4]
```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
...
+    slide,
+    onSlideChange,
+    onMount,
+    setAdvance,
     ...rest
   } = props;
   const childrenArray = toChildArray(children);
   const {length} = childrenArray;
-  const [curSlide, setCurSlide] = useState(defaultSlide || 0);
+  const [curSlide, setCurSlide] = useState(slide ? slide : defaultSlide || 0);
```
Comment (jridgewell) [inline extensions/amp-base-carousel/1.0/base-carousel.js:41]: Can you explain this change? This is the default value, so `slide` shouldn't matter? What is `slide` supposed to do?

File: extensions/amp-base-carousel/1.0/amp-base-carousel.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+/** @override */
+AmpBaseCarousel['Component'] = BaseCarousel;
+
+/** @override */
+AmpBaseCarousel['children'] = {
+  'children': {
+    name: 'children',
+    selector: ':not([slot="prev-arrow"]):not([slot="next-arrow"])',
```
Comment (jridgewell) [inline extensions/amp-base-carousel/1.0/amp-base-carousel.js:115]: I don't think we actually need this selector. `children` catch-all should take lower priority than the other defined children nodes.

File: extensions/amp-base-carousel/1.0/base-carousel.js (hunk 4 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.js:41]: `slide` and `defaultSlide` are supposed to parallel `value` and `defaultValue` for controlled versus uncontrolled inputs. This particular case is to allow `amp-base-carousel` to mutate the `slide` prop in order to display a different slide. This is helpful for the `goToSlide` action.

I hear you in the next line there is `slideState` which also considers them both. I don't feel strongly about this change and can revert it.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.js (hunk 5 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.js]: Changed to `*`

File: extensions/amp-base-carousel/1.0/base-carousel.js [hunk 6]
```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
...
+    slide,
+    onSlideChange,
+    onMount,
+    setAdvance,
     ...rest
   } = props;
   const childrenArray = toChildArray(children);
   const {length} = childrenArray;
   const [curSlide, setCurSlide] = useState(defaultSlide || 0);
+  const slideState = slide ? slide : curSlide;
```
Comment (jridgewell) [inline extensions/amp-base-carousel/1.0/base-carousel.js:42]: This will have a bug if `slide` is `0`. Try:
```suggestion
  const slideState = slide ?? curSlide;
```

File: extensions/amp-base-carousel/1.0/base-carousel.js (hunk 4 above)
Comment (jridgewell) [reply extensions/amp-base-carousel/1.0/base-carousel.js]: Ok, this is because the component can either be controlled or uncontrolled. If I remember correctly, we had a helper from the React Training that would make this a bit easier…

Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.js]: I recall the discussion but not the resolution unfortunately 😅  @dvoytenko @samouri any ideas?

**Comment by caroqliu** [comment]
@dvoytenko @jridgewell addressed review comments and added unit tests, PTAL

File: build-system/eslint-rules/dict-string-keys.js [hunk 7]
```diff
@@ -42,7 +42,7 @@ function checkNode(node, context) {
     node.properties.forEach(function (prop) {
       if (!prop.key.raw && !prop.computed) {
         context.report({
-          node,
+          node: prop,
```
Comment (dvoytenko) [inline build-system/eslint-rules/dict-string-keys.js:45]: Does this belong in a separate pull request?

File: extensions/amp-base-carousel/1.0/base-carousel.js (hunk 4 above)
Comment (dvoytenko) [reply extensions/amp-base-carousel/1.0/base-carousel.js]: In controlled mode, there are some expectations:

1. A prop (e.g. `currentSlide`) that controls the parent slide is passed. The carousel can only display that slide.
2. A prop callback (e.g. `onSlideChange`) is called with the new intended slide. The caller updates `currentSlide` prop if it decides to update the carousel. Thus the even handler is critical to the carousel's functioning.
3. Notice that the state is not really needed here.
4. There's no need for public API to inquire the current slide, because the caller already knows it - it's the `currentSlide` prop.

In a non-controlled mode, the expectations are different:

1. A prop (e.g. `defaultSlide`) can control the initially open slide.
2. There's an internal state that tracks the current slide. It goes out of sync with the `defaultSlide`.
3. We'd still likely to have `onSlideChange` - but it's just an event with no expectations attached. It's not critical to the carousel's functioning.
4. We may need an imperative API so that caller could find out the current slide and change it imperatively.

This is something we tried in the prototype, but looking again at this, I'm not sure if the controlled mode is working well here. Although maybe we can find a compromise. But consider this: a user swipes and the carousel is scrolling gradually to the next slide. At some point the controlled `onSlideChange` is called (currently once we are 50% there). Here's what happens:
* If the caller wants to advance to the next slide and updates the prop - the scrolling transition is kind of in the way. If we preserve it - that's a lot of work. If we break it - the effect is very jarring.
* If the caller doesn't want to advance - the user experience is also very jarring - the carousel will jump back.

So, at this point, I'm not sure the controlled mode is such a good idea. I also don't think that opening a carousel to a non-first slide is a useful feature either. It definitely hurts with SSR. And sooner or later we have to look into imperative API anyway.

This is a long way to say that I think we should rip out the `slide` and `defaultSlide` props and controlled mode in general. We can keep `onSlideChange` however.

File: extensions/amp-base-carousel/1.0/base-carousel.type.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * @typedef {{
+ *   arrowNext: (?PreactDef.VNode|undefined),
+ *   arrowPrev: (?PreactDef.VNode|undefined),
+ *   children: (!PreactDef.Renderable),
+ *   defaultSlide: (number|undefined),
+ *   loop: (boolean|undefined),
+ *   slide: (number|undefined),
+ *   onMount: (function():undefined|undefined),
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.type.js:30]: This shouldn't be a public property. IMHO either let's make it private or remove.

File: extensions/amp-base-carousel/1.0/base-carousel.type.js [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+ *   onSlideChange: (function(number):undefined|undefined),
+ *   setAdvance: (function(function(number):undefined):undefined|undefined),
+ * }}
+ */
+BaseCarouselDef.Props;
+
+/**
+ * @typedef {{
+ *   children: !Array<PreactDef.Renderable>,
+ *   ignoreProgrammaticScroll: {current: (boolean|null)},
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.type.js:40]: Is this a ref? I'd prefer we suffix refs with `Ref` since it's a pretty internal thing per a component most of the time.

File: extensions/amp-base-carousel/1.0/base-carousel.type.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * @typedef {{
+ *   children: !Array<PreactDef.Renderable>
+ * }}
+ */
+BaseCarouselDef.SlideProps;
+
+/**
+ * @typedef {{
+ *   advance: (function():undefined|undefined),
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.type.js:57]: Is this a callback? Should it be `onAdvance`?

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+amp-base-carousel {
+  display: block;
+  overflow: hidden;
+}
+
+amp-base-carousel > * {
+  box-sizing: border-box !important;
+  margin: 0 !important;
+  flex-shrink: 0 !important;
+  height: 100%;
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/amp-base-carousel.css:26]: Hmm. This is a tricky one. IIUC, we layout the children here on AMP layer. But it's a critical part of the React component too? I.e. does the React component work correctly without these styles? 

I think there are two models here: 
1. Stretch: force children to take 100% space. Requires (more or less) stylesheet with `!important`.
2. Contain: have an internal slide wrapper that takes 100% of space, but uses centering flexbox for each slide's content.

Do we have any idea which one would work better from the component user point of view?

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+amp-base-carousel {
+  display: block;
+  overflow: hidden;
+}
+
+amp-base-carousel > * {
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/amp-base-carousel.css:22]: This might break service tags, no? E.g. placeholder, fallback, and, especially, sizer?

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  box-sizing: border-box !important;
+  margin: 0 !important;
+  flex-shrink: 0 !important;
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/amp-base-carousel.css:25]: This assumes that parent should be flex. But it's not in the above declaration?

File: build-system/eslint-rules/dict-string-keys.js (hunk 7 above)
Comment (caroqliu) [reply build-system/eslint-rules/dict-string-keys.js:45]: If that's preferred -- @jridgewell recently committed it to this PR because it is required for this to be able to pass the linter.

File: extensions/amp-base-carousel/1.0/base-carousel.type.js (hunk 8 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.type.js:30]: I can make it private for now. For my learning, why don't we want publishers to be able to access a common `onMount` hook on all Preact components?

File: extensions/amp-base-carousel/1.0/base-carousel.type.js (hunk 10 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.type.js:57]: No, it is a function that takes a slide distance and shows that slide. i.e. If we have slides `[1][2][3][4][5]` and are seeing slide `[1]`, then `advance(2)` will show slide `[3]`

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css (hunk 11 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.css:26]: Yes, I had thought about putting this in the Preact layer but left it in AMP for the following reasons:
- These are styles on the user-supplied DOM. In Preact we would have to merge styles if we want to insert them.
- Just like we leave AMP runtime to handle layout, we leave Preact publishers to specify the dimensions that they want their `BaseCarousel` and underlying slides -- IMO it's not good devX if we make them do this and only sometimes respect it. 

The original reason I ended up including these styles are to keep amp-base-carousel visually consistent between 0.1 and 1.0 for [this sample](https://github.com/ampproject/amphtml/blob/c14b507eda631475233d739e4653eb0c6009d101/test/manual/amp-base-carousel/captions.amp.html) involving captioned responsive images. I added the Preact counterpart sample in the Storybook if you want to take a look at these.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css (hunk 12 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.css:22]: Replaced `*` with a class name applied to the component's `realChildNodes`

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css (hunk 13 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.css:25]: This is an interesting one. For the slotted elements the [Preact parent will be flex.](https://github.com/ampproject/amphtml/blob/master/extensions/amp-base-carousel/1.0/base-carousel.css.js#L22)

File: extensions/amp-base-carousel/1.0/base-carousel.type.js (hunk 8 above)
Comment (dvoytenko) [reply extensions/amp-base-carousel/1.0/base-carousel.type.js]: I don't think I see this a lot. In theory, when I do `<MyComponent>` it's always mounted. So it'd seem like it'd be a rare need, at the very least.

Comment (dvoytenko) [reply extensions/amp-base-carousel/1.0/base-carousel.type.js]: Also, I'm going to start replacing owner system with context props. So I think this will become unneeded soon anyway.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css (hunk 13 above)
Comment (dvoytenko) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.css:25]: This deserves a comment, IMHO.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.js [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        fireSlideChangeEvent(
+          this.win,
+          action,
+          element,
+          index,
+          ActionTrust.HIGH
+        );
+        this.mutateProps(dict({'slide': index}));
+      },
+      'setAdvance': (a) => (advance = a),
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/amp-base-carousel.js:94]: We need to switch to `useImperativeHandle` here.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.js [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+ * Triggers a 'slideChange' event with one data param:
+ * 'index' - index of the current slide.
+ * @param {!Window} win
+ * @param {!../../../src/service/action-impl.ActionService} action
+ * @param {!Element} el The element that was selected or deslected.
+ * @param {number} index
+ * @param {!ActionTrust} trust
+ * @private
+ */
+function fireSlideChangeEvent(win, action, el, index, trust) {
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/amp-base-carousel.js:41]: Nit: place after `AmpBaseCarousel` and related declarations.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.js [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    dict({'index': index})
+  );
+  action.trigger(el, name, slideChangeEvent, trust);
+}
+
+class AmpBaseCarousel extends PreactBaseElement {
+  /** @override */
+  init() {
+    const {element} = this;
+    const action = Services.actionServiceForDoc(element);
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/amp-base-carousel.js:55]: Can we resolve this in the `fireSlideChangeEvent`?

File: extensions/amp-base-carousel/1.0/base-carousel.js [hunk 17]
```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
...
  * @return {PreactDef.Renderable}
  */
 export function BaseCarousel(props) {
   const {
-    'arrowPrev': arrowPrev,
-    'arrowNext': arrowNext,
-    'children': children,
-    'defaultSlide': defaultSlide,
-    'loop': loop,
+    arrowPrev,
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.js:28]: Why not destructure all these in the function argument?

File: extensions/amp-base-carousel/1.0/base-carousel.js [hunk 18]
```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
...
     // Modify scrollLeft is preferred to `setCurSlide` to enable smooth scroll.
     // Note: `setCurSlide` will still be called on debounce by scroll handler.
     container./* OK */ scrollLeft += container./* OK */ offsetWidth * dir;
   };
+  useMountEffect(() => {
+    if (onMount_) {
+      onMount_();
+    }
+    if (setAdvance) {
+      setAdvance(advance);
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.js:55]: Ditto: IMHO this should be forwardRef + `useImperativeHandle`.

File: extensions/amp-base-carousel/1.0/base-carousel.js [hunk 19]
```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
...
   const childrenArray = toChildArray(children);
   const {length} = childrenArray;
   const [curSlide, setCurSlide] = useState(defaultSlide || 0);
-  const ignoreProgrammaticScroll = useRef(true);
-  const setRestingIndex = (i) => {
-    ignoreProgrammaticScroll.current = true;
-    setCurSlide(i);
-  };
-  const scrollRef = useRef(null);
+  const slideState = slide ?? curSlide;
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.js:42]: I'd look into the issue of controlled/uncontrolled together with `useImpertiveHandle`. I recommend we rip out this code for now and add it in a separate PR. This PR could be dedicated to layout/scroller structure, which is already quite a good set of functionality.

File: extensions/amp-base-carousel/1.0/base-carousel.js [hunk 20]
```diff
@@ -17,43 +17,61 @@ import * as Preact from '../../../src/preact';
...
+    if (onMount_) {
+      onMount_();
+    }
+    if (setAdvance) {
+      setAdvance(advance);
+    }
+  });
+
+  const setRestingIndex = (i) => {
+    ignoreProgrammaticScrollRef.current = true;
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/base-carousel.js:60]: It looks like `ignoreProgrammaticScrollRef` is only used here and inside the `Scroller`. And the `setRestingIndex` is also only used inside the `Scroller`. I think that means that you can safely move `ignoreProgrammaticScrollRef` to the `Scroller`, which would reduce amount of state shared.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css (hunk 11 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.css:26]: Discussed offline: Placing styles in Preact rendered `<style>` tag. They must also stay in the AMP css file because the styles rendered in Preact do not propagate across the `<slot>` element in AMP mode.

File: extensions/amp-base-carousel/1.0/base-carousel.type.js (hunk 8 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.type.js]: In that case I will remove this `onMount` and the section where `amp-base-carousel` opts into the owners system to allow the AMP runtime to do this for now. I will also hold off on any `ChildLayoutManager` work for now.

File: extensions/amp-base-carousel/1.0/base-carousel.js (hunk 20 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.js:60]: Done.

File: extensions/amp-base-carousel/1.0/base-carousel.js (hunk 17 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/base-carousel.js:28]: Done.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.js (hunk 16 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.js:55]: Done.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.js (hunk 15 above)
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.js:41]: Done.

**Comment by caroqliu** [comment]
All comments addressed -- `useImperativeHandle` changes will be moved in a subsequent PR. For now the uncontrolled/controlled input pattern has been removed, along with it the `slides`, `defaultSlides` props and the `goToSlide` AMP action.

File: extensions/amp-base-carousel/1.0/scroller.js [hunk 21]
```diff
@@ -76,13 +71,13 @@ export function Scroller(props) {
...
     const container = (scrollRef.current = containerRef.current);
-    ignoreProgrammaticScroll.current = true;
+    ignoreProgrammaticScrollRef.current = true;
     setStyle(container, 'scrollBehavior', 'auto');
     container./* OK */ scrollLeft = loop
       ? container./* OK */ offsetWidth * pivotIndex
       : container./* OK */ offsetWidth * restingIndex;
     setStyle(container, 'scrollBehavior', 'smooth');
-  }, [ignoreProgrammaticScroll, loop, restingIndex, pivotIndex, scrollRef]);
+  }, [ignoreProgrammaticScrollRef, loop, restingIndex, pivotIndex, scrollRef]);
```
Comment (dvoytenko) [inline extensions/amp-base-carousel/1.0/scroller.js:80]: `ref` doesn't need to be in the deps now since it's defined up top.

Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/scroller.js:80]: Good catch, thank you.

File: extensions/amp-base-carousel/1.0/amp-base-carousel.css [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+amp-base-carousel {
+  display: block;
+  overflow: hidden;
+}
+
+.i-amphtml-carousel-slide {
```
[the remark this answers is not in the document]
Comment (caroqliu) [reply extensions/amp-base-carousel/1.0/amp-base-carousel.css:22]: Removed
